src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,297 @@
+            if (_item == null)
+            {
+                throw Error.Argument("item", SRResources.ChangedObjectTypeMismatch, typeof(TStructuralType), item.GetType());
+            }
+
+            _items.Add(_item);
+            base.InsertItem(index, item);
+        }
+
+        /// <inheritdoc/>
+        protected override void ClearItems()
+        {
+            _items.Clear();
+        }
+
+        /// <inheritdoc/>
+        protected override void RemoveItem(int index)
+        {
+            Contract.Assert(_items.Count > index);
+            
+            _items.RemoveAt(index);
+            base.RemoveItem(index);
+        }
+
+        /// <inheritdoc/>
+        protected override void SetItem(int index, IEdmChangedObject item)
+        {
+            IEdmChangedObject<TStructuralType> _item = item as IEdmChangedObject<TStructuralType>;
+            Contract.Assert(_item != null);
+
+            _items[index] = _item;
+        }
+
+        /// <summary>
+        /// Copy changed values is an implementation of Patch
+        /// </summary>
+        /// <param name="original">Original collection of the Type which needs to be updated</param>              
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+        internal EdmChangedObjectCollection<TStructuralType> CopyChangedValues(ICollection<TStructuralType> original)
+        {
+            //Here we need to Find the key of the Type, then only we will be able to find from the collection that which item in the collection
+            //corresponds to the item in delta list(Edmchangedobjectcoll).
+            Type type = typeof(TStructuralType);
+            IEnumerable<IEdmStructuralProperty> keys = EntityType.Key();
+
+            EdmChangedObjectCollection<TStructuralType> edmChangedObjectCollection = new EdmChangedObjectCollection<TStructuralType>(EntityType);
+
+            foreach (dynamic changedObj in _items)
+            {
+                DataModificationOperationKind operation = DataModificationOperationKind.Update;
+                IEdmChangedObject edmChangedObject = null;
+               
+                //Get filtered item based on keys
+                TStructuralType originalObj = GetFilteredItem(type, keys, original, changedObj); ;
+
+                try
+                {
+                    IEdmDeltaDeletedEntityObject deletedObj = changedObj as IEdmDeltaDeletedEntityObject;
+
+                    if (deletedObj != null)
+                    {
+                        if (deletedObj.Reason == DeltaDeletedEntryReason.Deleted)
+                        {
+                            operation = DataModificationOperationKind.Delete;
+                        }
+                        else
+                        {
+                            operation = DataModificationOperationKind.Unlink;
+                        }
```

Comment:
deletedObj.Reason may be null, in which case we should assume the operation was a delete -- we just don't know what kind of delete. So we should special case deletedObj.Reason == Removed as DataModificationOperationKind.Unlink, and for anything else return DataModificationOperationKind.Delete.  i.e, reverse your if/else; 
~~~c#
                        if (deletedObj.Reason == DeltaDeletedEntryReason.Removed)
                        {
                            operation = DataModificationOperationKind.Unlink;
                        }
                        else
                        {
                            operation = DataModificationOperationKind.Delete;
                        }
~~~